File: annotationprocessortesting/src/main/java/com/tickaroo/tikxml/regressiontests/paths/Company.java
package com.tickaroo.tikxml.regressiontests.paths;

import com.tickaroo.tikxml.annotation.Element;
import com.tickaroo.tikxml.annotation.ElementNameMatcher;
import com.tickaroo.tikxml.annotation.GenericAdapter;
import com.tickaroo.tikxml.annotation.Path;
import com.tickaroo.tikxml.annotation.Xml;
import com.tickaroo.tikxml.regressiontests.paths.element.Book;
import java.util.List;
import java.util.Objects;

/**
 * @author Hannes Dorfmann
 */
@Xml
public class Company {

  @Path("department/persons")
  @Element
  public List<Person> persons;

  @Element
  public List<Person> bosses;

  @Element
  public List<Statisch> statisches;

  @Override public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof Company)) return false;

    Company company = (Company) o;

    return Objects.equals(persons, company.persons);
  }

  @Override public int hashCode() {
    return persons != null ? persons.hashCode() : 0;
  }

  @GenericAdapter
  interface Statisch{}

  @Xml
  static class TestStatisch implements Statisch{}
}

File: annotationprocessortesting-kt/src/main/java/com/tickaroo/tikxml/regressiontests/paths/Company.kt
package com.tickaroo.tikxml.regressiontests.paths

import com.tickaroo.tikxml.annotation.Element
import com.tickaroo.tikxml.annotation.ElementNameMatcher
import com.tickaroo.tikxml.annotation.Path
import com.tickaroo.tikxml.annotation.Xml

/**
 * @author Hannes Dorfmann
 */
@Xml
class Company {

    @Path("department/persons")
    @Element(typesByElement = [
        ElementNameMatcher(type = Person::class),
        ElementNameMatcher(type = Employee::class),
        ElementNameMatcher(type = Boss::class)
    ])
    var persons: List<@JvmSuppressWildcards Person>? = null

    @Path("department/emptyPersons")
    @Element(typesByElement = [
        ElementNameMatcher(type = Person::class, name="person1"),
        ElementNameMatcher(type = Employee::class),
        ElementNameMatcher(type = Boss::class)
    ])
    var emptyPersons: List<@JvmSuppressWildcards Person>? = null

    @Path("department") @Element
    var room: Room? = null

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Company) return false

        val company = other as Company?

        return if (persons != null) persons == company!!.persons else company!!.persons == null
    }

    override fun hashCode(): Int {
        return if (persons != null) persons!!.hashCode() else 0
    }
}
